ddprof-lib/src/main/cpp/reservoirSampler.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2024 Datadog
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef RESERVOIR_SAMPLER_H
+#define RESERVOIR_SAMPLER_H
+
+#include <math.h>
+#include <random>
+#include <vector>
+
+
+template <class T>
+class ReservoirSampler {
+private:
+    const int size;
+    std::mt19937 generator;
+    std::uniform_real_distribution<double> uniform;
+    std::uniform_int_distribution<int> random_index;
+    std::vector<T> reservoir;
+
+public:
+    ReservoirSampler(const int size) :
+        size(size),
+        generator((std::random_device())()),
+        uniform(1e-16, 1.0),
+        random_index(0, size - 1) {
+        reservoir.reserve(size);
+    }
+
+    std::vector<T>& sample(const std::vector<T> &input) {
```

Comment:
I won't go into the actual algorithm we chose here and whether it was the correct one. Please tell me if I should spend more time here.

Reply:
I am quite confident of the algorithm here. 
Probably not worth spending a lot of time on it - @richardstartin has a link to an article describing this algorithm, IIRC, if you are interested.